src/ui/ui_transaction_nbgl.c

```diff
@@ -0,0 +1,230 @@
+#include <string.h>   // memset
+
+#include "os.h"
+#include "os_io_seproxyhal.h"
+#include "ux.h"
+#include "nbgl_use_case.h"
+
+#include "ui_main_menu.h"
+#include "ui_transaction.h"
+#include "ui_shared.h"
+#include "btchip_internal.h"
+#include "glyphs.h"
+
+typedef enum {
+    TX_TYPE_SINGLE_OUTPUT = 0,
+    TX_TYPE_FULL_REVIEW,
+    TX_TYPE_FINALIZE,
+    TX_TYPE_SIGN_MESSAGE,
+} transaction_type_t;
+
+typedef struct {
+    const char* reviewStart;
+    const char* reviewCancel;
+    const char* finishOk;
+    const char* finishCancel;
+    const char* choiceCancel;
+    const char* choiceGoBack;
+} messages_t;
+
+static nbgl_layoutTagValueList_t pairList;
+static nbgl_layoutTagValue_t pairs[4];
+static nbgl_pageInfoLongPress_t infoLongPress;
+static transaction_type_t txType;
+static messages_t msgs;
+static char genericText[70];
+static bool reviewStarted;
+
+static void reviewChoice(bool confirm);
+static void rejectChoice(bool confirm);
+static void rejectUseCaseChoice(void);
+static void displayTransaction(void);
+static void reviewStart(void);
+
+static void rejectChoice(bool confirm) {
+    if (confirm) {
+        nbgl_useCaseStatus(msgs.finishCancel, true, ui_idle);
+        txType == TX_TYPE_SIGN_MESSAGE ? io_seproxyhal_touch_message_signature_verify_cancel(NULL)
+                                       : io_seproxyhal_touch_verify_cancel(NULL);
+    } else {
+        if (reviewStarted) {
+            displayTransaction();
+        } else {
+            reviewStart();
+        }
+    }
+}
+
+static void reviewChoice(bool confirm) {
+    if (confirm) {
+        if (txType != TX_TYPE_SINGLE_OUTPUT) {
+            nbgl_useCaseStatus(msgs.finishOk, true, ui_idle);
+        }
+        txType == TX_TYPE_SIGN_MESSAGE ? io_seproxyhal_touch_message_signature_verify_ok(NULL)
+                                       : io_seproxyhal_touch_verify_ok(NULL);
+    } else {
+        rejectUseCaseChoice();
+    }
+}
+
+static void rejectUseCaseChoice(void) {
+    nbgl_useCaseChoice(NULL,
+                       msgs.choiceCancel,
+                       NULL,
+                       "Yes, cancel",
+                       msgs.choiceGoBack,
+                       rejectChoice);
+}
+
+static void displayTransaction(void) {
+    reviewStarted = true;
+    nbgl_useCaseStaticReview(&pairList, &infoLongPress, msgs.reviewCancel, reviewChoice);
+}
+
+static void reviewStart(void) {
+    explicit_bzero(pairs, sizeof(pairs));
+    explicit_bzero(&infoLongPress, sizeof(infoLongPress));
+
+    msgs.reviewStart = "Review\ntransaction";
+    msgs.reviewCancel = "Cancel transaction";
+    msgs.choiceCancel = "Cancel transaction ?";
+    msgs.choiceGoBack = "Go back to transaction";
+    msgs.finishOk = "TRANSACTION\nSIGNED";
+    msgs.finishCancel = "Transaction rejected";
+
+    infoLongPress.text = "Accept and send";
+    infoLongPress.longPressText = "Hold to accept";
+    infoLongPress.icon = &C_decred_icon_64px;
+
+    pairs[0].item = "Amount";
```

Comment:
I dont understand why you are filling the pairs here and then overiding them in the scwitch case

Reply:
Same for reviewStart for example

Reply:
I have modified it to leave only common strings set before the switch